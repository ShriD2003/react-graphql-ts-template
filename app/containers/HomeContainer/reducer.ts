import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';
import { Launch } from '@app/containers/HomeContainer';

export type HomeContainerState = {
  launchQuery: string;
  launchData: {
    launches?: Launch[];
  };
  launchListError: any;
  loading: boolean;
};

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetLaunchList: ['launchQuery'],
  successGetLaunchList: ['launchData'],
  failureGetLaunchList: ['launchListError']
});

export const initialState: HomeContainerState = {
  loading: false,
  launchQuery: '',
  launchData: {},
  launchListError: null
};

export interface HomeContainerAction extends Partial<Omit<HomeContainerState, 'loading'>> {
  type?: string;
}

export const homeContainerReducer = (state = initialState, action: HomeContainerAction) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_LAUNCH_LIST:
        draft.loading = true;
        draft.launchQuery = action.launchQuery || '';
        break;

      case homeContainerTypes.SUCCESS_GET_LAUNCH_LIST:
        draft.launchData = action.launchData!;
        draft.loading = false;
        break;
      case homeContainerTypes.FAILURE_GET_LAUNCH_LIST:
        draft.loading = false;
        draft.launchListError = draft.launchListError = get(action.launchListError, 'message', 'something_went_wrong');
        break;
    }
  });

export default homeContainerReducer;
