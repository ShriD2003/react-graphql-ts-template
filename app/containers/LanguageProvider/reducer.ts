/*
 *
 * LanguageProvider reducer
 *
 */
import { createActions } from 'reduxsauce';
import produce from 'immer';
import { DEFAULT_LOCALE } from '@app/i18n';

type LanguageProviderTypes = {
  CHANGE_LOCALE: 'CHANGE_LOCALE';
};

type LanguageProviderAction = {
  type?: keyof LanguageProviderTypes;
  locale?: string;
};

type LanguageProviderCreators = {
  changeLocale: (locale: 'en') => LanguageProviderAction;
};

export const { Types: languageProviderTypes, Creators: languageProviderActions } = createActions<
  LanguageProviderTypes,
  LanguageProviderCreators
>({
  changeLocale: ['locale']
});

export const initialState = {
  locale: DEFAULT_LOCALE
};

/* eslint-disable default-case, no-param-reassign */
export const languageProviderReducer = (state = initialState, action: LanguageProviderAction) =>
  produce(state, (draft) => {
    switch (action.type) {
      case languageProviderTypes.CHANGE_LOCALE:
        if (action.locale) {
          draft.locale = action.locale;
        }
        break;
    }
  });

export default languageProviderReducer;
