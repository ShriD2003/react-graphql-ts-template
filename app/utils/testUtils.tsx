import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
