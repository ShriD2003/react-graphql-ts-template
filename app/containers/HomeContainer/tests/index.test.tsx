import React from 'react';
import { timeout, renderProvider } from '@utils/testUtils';
import { HomeContainerProps, HomeContainerTest as HomeContainer, LaunchData, mapDispatchToProps } from '../index';
import { homeContainerTypes } from '../reducer';

describe('<HomeContainer /> tests', () => {
  let submitSpy: jest.Mock;
  let defaultProps: HomeContainerProps;

  beforeEach(() => {
    submitSpy = jest.fn();
    defaultProps = {
      loading: true,
      dispatchLaunchList: submitSpy,
      launchQuery: ''
    };
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<HomeContainer {...defaultProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchLaunchList on page reload', async () => {
    renderProvider(<HomeContainer {...defaultProps} />);
    await timeout(500);
    expect(submitSpy).toBeCalledWith();
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchLaunchListSpy = jest.fn();
    const actions = {
      dispatchLaunchList: { type: homeContainerTypes.REQUEST_GET_LAUNCH_LIST },
      dispatchClearLaunchList: { type: homeContainerTypes.CLEAR_LAUNCH_LIST }
    };

    const props = mapDispatchToProps(dispatchLaunchListSpy);
    props.dispatchLaunchList();
    expect(dispatchLaunchListSpy).toHaveBeenCalledWith(actions.dispatchLaunchList);
  });

  it('should render default error message when search goes wrong', () => {
    const defaultError = 'This is an error';
    const { getByTestId } = renderProvider(
      <HomeContainer {...defaultProps} launchListError={defaultError} loading={false} />
    );
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-card')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });

  it('should render the data when loading becomes false', () => {
    const launchData: LaunchData = {
      launches: [
        {
          missionName: 'Sample Mission',
          launchDateLocal: 'some date',
          links: {
            wikipedia: 'sample link',
            flickrImages: ['sample image']
          }
        }
      ]
    };
    const { getByTestId } = renderProvider(<HomeContainer launchData={launchData} {...defaultProps} />);
    expect(getByTestId('list')).toBeInTheDocument();
  });

  it('should render Skeleton Comp when "loading" is true', async () => {
    const { baseElement } = renderProvider(<HomeContainer {...defaultProps} loading={true} />);

    expect(baseElement.getElementsByClassName('ant-skeleton').length).toBe(1);
  });
});
