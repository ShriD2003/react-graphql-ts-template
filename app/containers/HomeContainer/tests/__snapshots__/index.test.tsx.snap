// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeContainer /> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="HomeContainer__Container-wd79o1-0 dbhtxz"
    >
      <div
        class="HomeContainer__CustomHeader-wd79o1-1 hlkaQK"
      >
        <span
          class="ant-input-group-wrapper ant-input-search"
        >
          <span
            class="ant-input-wrapper ant-input-group"
          >
            <input
              class="ant-input"
              data-testid="search-bar"
              type="text"
              value=""
            />
            <span
              class="ant-input-group-addon"
            >
              <button
                class="ant-btn ant-btn-icon-only ant-input-search-button"
                type="button"
              >
                <span
                  aria-label="search"
                  class="anticon anticon-search"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="search"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </span>
        <button
          class="ant-btn"
          disabled=""
          type="button"
        >
          <span>
            CLEAR SORT
          </span>
        </button>
        <div
          class="ant-select HomeContainer__SortSelect-wd79o1-2 gXPPUN ant-select-single ant-select-allow-clear ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-search"
            >
              <input
                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autocomplete="off"
                class="ant-select-selection-search-input"
                id="rc_select_TEST_OR_SSR"
                readonly=""
                role="combobox"
                style="opacity: 0;"
                type="search"
                unselectable="on"
                value=""
              />
            </span>
            <span
              class="ant-select-selection-item"
              title="SORT DATE"
            >
              SORT DATE
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          >
            <img
              alt="chevron-up-down"
              src="IMAGE_MOCK"
            />
          </span>
          <span
            aria-hidden="true"
            class="ant-select-clear"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              aria-label="close-circle"
              class="anticon anticon-close-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
      <div
        data-testid="list"
      >
        <div
          class="ant-skeleton ant-skeleton-active"
        >
          <div
            class="ant-skeleton-content"
          >
            <h3
              class="ant-skeleton-title"
              style="width: 38%;"
            />
            <ul
              class="ant-skeleton-paragraph"
            >
              <li />
              <li />
              <li
                style="width: 61%;"
              />
            </ul>
          </div>
        </div>
      </div>
      <div
        class="HomeContainer__CustomFooter-wd79o1-3 kMFGHV"
      >
        <button
          class="ant-btn ant-btn-primary"
          disabled=""
          type="button"
        >
          <span>
            PREV
          </span>
        </button>
        <button
          class="ant-btn ant-btn-primary"
          disabled=""
          type="button"
        >
          <span>
            NEXT
          </span>
        </button>
      </div>
    </div>
  </div>
</body>
`;
